### PR TITLE
:sparkles: Add types for all the public API

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
Related to #128

### CODE REVIEW

File: s3path.py
Comment (nlangellier) [inline s3path.py:758]: I'm curious to understand why you didn't type hint the output of this method and `from_bucket_key`. Naively I would have thought you could do
```python
@classmethod
def from_uri(cls, uri: str) -> "PureS3Path":
```
Is there some reason not to do this?

Comment (gabrieldemarmiesse) [reply s3path.py:758]: I'm just splitting the change in multiple PRs to avoid doing a big one,it's possible to do this indeed

File: s3path.py
Comment (nlangellier) [inline s3path.py:891]: I think for this one you want `from typing import Generator` and then change this to
```suggestion
    def iterdir(self) -> Generator["S3Path", None, None]:
```

File: s3path.py
Comment (nlangellier) [inline s3path.py:1046]:
```suggestion
    def samefile(self, other_path: Union[str, "S3Path"]) -> bool:
```

**Comment by gabrieldemarmiesse** [comment]
I like the `from __future__ import annotations` but unlike other `__future__` it's not 100% sure that it will be the default in the future. See https://docs.python.org/3/library/__future__.html#id2 . If it gets rejected, then we will have to change the whole codebase so I'd prefer to avoid using it.

File: s3path.py
Comment (gabrieldemarmiesse) [reply s3path.py:891]: Iterator is also allowed for functions with yield. This example comes from the python documentation itself:
```python
def infinite_stream(start: int) -> Iterator[int]:
    while True:
        yield start
        start += 1
```

https://docs.python.org/3/library/typing.html#typing.Generator

File: s3path.py
Comment (gabrieldemarmiesse) [reply s3path.py:899]: Same as above

File: s3path.py
Comment (gabrieldemarmiesse) [reply s3path.py:933]: Same as above

File: s3path.py
Comment (nlangellier) [reply s3path.py:891]: all this time i've been over-complicating the return types of my yield-only generators. thanks for the lesson!

**Comment by nlangellier** [comment]
> I like the `from __future__ import annotations` but unlike other `__future__` it's not 100% sure that it will be the default in the future. See https://docs.python.org/3/library/__future__.html#id2 . If it gets rejected, then we will have to change the whole codebase so I'd prefer to avoid using it.

Makes sense. I'm not strongly opinionated on this.

**Review by nlangellier — COMMENTED** [review]
LGTM, but as mentioned above, not going to approve as I am not a maintainer of this repo.

File: s3path.py
Comment (liormizr) [inline s3path.py:798]: Please add return type hint

**Review by liormizr — COMMENTED** [review]
Hi all
First I like to say thank you for the PR and great reviews

I agree with all of you and think that it will be better for future versions to add and us 
`from __future__ import annotations` now when we are adding types

Now about the Iterator VR Generator typing.
I think we need to be explicit as possible.
For a method with a `yield` statements a Generator type hint is more precise.

So to conclude my review
1. Lets add `from __future__ import annotations` and add return type accordingly
2. Lets change methods with `yield` statements return type hint to Generator
3. @gabrieldemarmiesse I'm OK with splitting this task to multiple PR's, but if we are changing method signature lets do it end to end so we won't need to remember it in the next PR.
4. How do you recommend changing the doc's? or not changing the doc's?

I added some comments as well but this comment summaries all me feedback

Thanks again!

**Comment by gabrieldemarmiesse** [comment]
Done @liormizr  could I have another code review? *I encourage your to use squash and merge est the merging option* because otherwise master will be very messy, there are lots of commits in my pull request.

**Comment by gabrieldemarmiesse** [comment]
hi @liormizr  I had some time this weekend and typed the whole public api. I also fixed some return types with Self that will now return the correct type if there is some subclassing (otherwise the type is not narrowed down, see https://peps.python.org/pep-0673/ ).

**Comment by liormizr** [comment]
@gabrieldemarmiesse I added to master a fix for the issue that broke the tests
Can you please merge from master so we will rerun the tests?

Thanks

File: s3path.py
Comment (liormizr) [inline s3path.py:984]: spacing is off on the buffering arg line

File: s3path.py
Comment (liormizr) [inline s3path.py:913]: same comment here on the self type hint

File: s3path.py
Comment (liormizr) [inline s3path.py:921]: same comment here on the self type hint
You get me :-)
I wont write it for all the rest of the places

File: s3path.py
Comment (gabrieldemarmiesse) [inline s3path.py:984]:
```suggestion
            buffering: int = DEFAULT_BUFFER_SIZE,
```

File: s3path.py
Comment (liormizr) [reply s3path.py:780]: I find type hints useful in two cases: documentation and helping the IDE auto complete.
I don’t plan to use type checkers such as mypy or pyright at this time.

I’d like to keep the code as simple as possible, and that includes the amount a developer has to read.

I’d like to add type hints, but we should do it mostly in order to help with documentation and maybe help the IDE if there are proven places where it fails to autocomplete.

So at this time I prefer not to add this level of type hints.

@gabrieldemarmiesse it's your call what to do know.
You can continue or I can create a new PR and work on it.

I apologize for rejecting the PR as it's know, but the work is not lost - we can use parts of it, and of course I learned a lot in the process.

Comment (gabrieldemarmiesse) [reply s3path.py:780]: No worries, I can change back and remove the Self type, let me do that

**Comment by gabrieldemarmiesse** [comment]
@liormizr can you re-review?

**Comment by gabrieldemarmiesse** [comment]
@liormizr is there something I can do to push this forward? I think I've adressed all the issues you mentionned.

**Review by liormizr — APPROVED** [review]

**Pull request merged**
